### PR TITLE
[htlcswitch] Optionally force close channel on link failure

### PR DESCRIPTION
**Pull request by halseth**
This PR adds the option to force close a channel in case of a failure within the link. Now calls to `fail(...)` will always mark the channel inactive and disconnect the peer, and can optionally force close the channel.

### TODO:
- [x] determine which cases we want to force close the channel for sure. Some cases are possible to recover from, and we should only force close if we absolutely _have to_.
- [x] expand unit test after decision is made on the above.
- [x] integration test?

Putting PR up to get some initial feedback.

### CODE REVIEW

**Comment by Roasbeef** [comment]
>  determine which cases we want to force close the channel for sure. Some cases are possible to recover from, and we should only force close if we absolutely have to,

I'd say we only want to force close if: the remote party sends us an invalid commitment (we can't verify signatures) or the remote party violates any of the commitment constraints (reserve, etc as atm we can't recover from this). So only cases that we can't recover from assuming that the remote party will mechanically rebroadcast the violating commitment.

File: htlcswitch/link.go [hunk 1]
```diff
@@ -761,7 +768,7 @@ func (l *channelLink) htlcManager() {
 	if l.cfg.SyncStates {
 		if err := l.syncChanStates(); err != nil {
 			l.errorf("unable to synchronize channel states: %v", err)
-			l.fail(err.Error())
+			l.fail(false, err.Error())
```
Comment (Roasbeef) [inline htlcswitch/link.go]: We'll need to revisit this after #1131 is complete.

File: htlcswitch/link.go [hunk 2]
```diff
@@ -1080,8 +1087,7 @@ func (l *channelLink) handleDownStreamPkt(pkt *htlcPacket, isReProcess bool) {
 			pkt.destRef,
 			&closedCircuitRef,
 		); err != nil {
-			// TODO(roasbeef): broadcast on-chain
-			l.fail("unable to settle incoming HTLC: %v", err)
+			l.fail(true, "unable to settle incoming HTLC: %v", err)
```
Comment (Roasbeef) [inline htlcswitch/link.go]: Possibly we only want to force close if we're unable to receive _their_ settle?

File: htlcswitch/link.go [hunk 3]
```diff
@@ -2151,15 +2154,15 @@ func (l *channelLink) processRemoteAdds(fwdPkg *channeldb.FwdPkg,
...
+				l.fail(true, "unable to settle htlc: %v", err)
 				return false
 			}
 
 			// Notify the invoiceRegistry of the invoices we just
 			// settled with this latest commitment update.
 			err = l.cfg.Registry.SettleInvoice(invoiceHash)
 			if err != nil {
-				l.fail("unable to settle invoice: %v", err)
+				l.fail(true, "unable to settle invoice: %v", err)
```
Comment (Roasbeef) [inline htlcswitch/link.go]: This might be a bit too aggressive as this may be an issue with our db state or the implementation of the interface.

File: htlcswitch/link.go [hunk 4]
```diff
@@ -2481,11 +2484,31 @@ func (l *channelLink) sendMalformedHTLCError(htlcIndex uint64,
 }
 
 // fail helper function which is used to encapsulate the action necessary for
-// proper disconnect.
-func (l *channelLink) fail(format string, a ...interface{}) {
+// proper disconnect. If forceClose is true, we will in addition to
+// disconnecting the peer unilateraly close the channel.
+func (l *channelLink) fail(forceClose bool, format string, a ...interface{}) {
+	// TODO(halseth): optionally send error to remote?
```
Comment (Roasbeef) [inline htlcswitch/link.go]: Ideally yes, with some details as to _why_ we failed the channel.

File: htlcswitch/link.go [hunk 5]
```diff
@@ -2481,11 +2484,31 @@ func (l *channelLink) sendMalformedHTLCError(htlcIndex uint64,
...
 	reason := errors.Errorf(format, a...)
-	log.Error(reason)
+	l.errorf("Failing link: %s", reason)
+
+	l.cfg.Peer.WipeChannel(l.channel.ChannelPoint())
```
Comment (Roasbeef) [inline htlcswitch/link.go]: IIRC, the `Disconnect` method should wipe the channel itself? Or even once the peer disconnects, then the `peerTerminationWatcher` will remove it from the link all together.

File: htlcswitch/link_test.go [hunk 6]
```diff
@@ -1494,8 +1494,8 @@ func newSingleLinkTestHarness(chanAmt, chanReserve btcutil.Amount) (
 
 	const startingHeight = 100
 	aliceLink := NewChannelLink(aliceCfg, aliceChannel, startingHeight)
-	if err := aliceSwitch.AddLink(aliceLink); err != nil {
-		return nil, nil, nil, nil, nil, err
+	start := func() error {
```
Comment (Roasbeef) [inline htlcswitch/link_test.go]: Why's this necessary?

Comment (halseth) [reply htlcswitch/link_test.go]: This is done such that we can later set up the link to fail when it is added to the switch. (in this case we use it to make `syncStates` fail at startup)

File: htlcswitch/link.go (hunk 5 above)
Comment (halseth) [reply htlcswitch/link.go]: Yep, removed it.

**Comment by halseth** [comment]
We currently only force close in _two cases_ now:
- When receiving an invalid commitment
- When receiving an invalid settle

In all other cases we will still fail the link, but we won't force close the channel. The rationale for this is to be a bit conservative to begin with, since the user can force close the channel manually if need be. In the two mentioned cases we force close, as it is more likely the remote is malicious. Feel free to disagree.

**Comment by Roasbeef** [comment]
Yeah I think those two are a starting point. In a sense, if there're any constraint related violations, those point to a cross-implementation bug which leaving around the channel for a bit of time would help us to investigate. Once we get some of this monitoring infrastructure up, should be easy for node operators to set up thresholds for this stuff so we can obtain logs soon after the occurrence.

File: htlcswitch/link.go [hunk 7]
```diff
@@ -33,14 +34,93 @@ const (
...
+	// ErrInvalidRevocation indicates that the remote peer send us an
+	// invalid revocation message.
+	ErrInvalidRevocation
+)
+
+// linkFailureError encapsulates an error that will make us fail the current
+// link. It contains the necessary information needed to determine if we should
+// force close the channel in the process, and if any error data should be sent
+// to the peer.
+type linkFailureError struct {
```
Comment (Roasbeef) [inline htlcswitch/link.go]: 👍

File: htlcswitch/link.go [hunk 8]
```diff
@@ -2503,12 +2610,56 @@ func (l *channelLink) sendMalformedHTLCError(htlcIndex uint64,
...
+
+	if linkErr.forceClose {
+		l.warnf("Force closing link.")
+
+		// If we have decided to force close this channel, we'll wait
+		// for the method to return with a close tx. This ensures that
+		// the channel arbitrator has properly removed the link from
+		// our set of active links, and broadcasted the closing tx,
+		// such that no new updates will be attempted on this channel.
+		closeTx, err := l.cfg.ForceClose()
```
Comment (Roasbeef) [inline htlcswitch/link.go]: In practice, I think this has the possibility to deadlock: we'll call into the `contractcourt`, which will then attempt to make a sync call to the switch to disconnect the link, however we're blocked on this initial call leading to a deadlock. As a result, we may need to run this asynchronously. It should be pretty straight forward to attempt to introduce the deadlock locally: open a channel, then shutdown one node and like change the csv value or something, then attempt to send a payment. The receiving node should then attempt to automatically force close.

**Comment by Roasbeef** [comment]
Confirmed locally the current format leads to a deadlock. 

We go to force close the channel:
```
1 @ 0x102ea9a 0x103e650 0x14fc9e3 0x1851ff0 0x166abde 0x16642d2 0x166131c 0x105b7c1
#	0x14fc9e2	github.com/lightningnetwork/lnd/contractcourt.(*ChainArbitrator).ForceCloseContract+0x452	/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/contractcourt/chain_arbitrator.go:633
#	0x1851fef	main.(*peer).loadActiveChannels.func2+0x5f							/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/peer.go:438
#	0x166abdd	github.com/lightningnetwork/lnd/htlcswitch.(*channelLink).fail+0x1ad				/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/htlcswitch/link.go:2656
#	0x16642d1	github.com/lightningnetwork/lnd/htlcswitch.(*channelLink).handleUpstreamMsg+0x1011		/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/htlcswitch/link.go:1375
#	0x166131b	github.com/lightningnetwork/lnd/htlcswitch.(*channelLink).htlcManager+0x8fb			/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/htlcswitch/link.go:1028
```

The the `channelAttendant` is block attempting to remove the link from the switch: 
```
1 @ 0x102ea9a 0x102eb4e 0x103f2d4 0x103efed 0x10634f8 0x10643fd 0x16789c2 0x185644c 0x1514698 0x15022e1 0x150425f 0x1509263 0x105b7c1
#	0x103efec	sync.runtime_SemacquireMutex+0x3c								/usr/local/go/src/runtime/sema.go:71
#	0x10634f7	sync.(*Mutex).Lock+0x107									/usr/local/go/src/sync/mutex.go:134
#	0x10643fc	sync.(*RWMutex).Lock+0x2c									/usr/local/go/src/sync/rwmutex.go:93
#	0x16789c1	github.com/lightningnetwork/lnd/htlcswitch.(*Switch).RemoveLink+0x41				/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/htlcswitch/switch.go:1820
#	0x185644b	main.newServer.func8+0x8b									/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/server.go:455
#	0x1514697	github.com/lightningnetwork/lnd/contractcourt.newActiveChannelArbitrator.func1+0x227		/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/contractcourt/chain_arbitrator.go:235
#	0x15022e0	github.com/lightningnetwork/lnd/contractcourt.(*ChannelArbitrator).stateStep+0x120		/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/contractcourt/channel_arbitrator.go:443
#	0x150425e	github.com/lightningnetwork/lnd/contractcourt.(*ChannelArbitrator).advanceState+0x17e		/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/contractcourt/channel_arbitrator.go:651
#	0x1509262	github.com/lightningnetwork/lnd/contractcourt.(*ChannelArbitrator).channelAttendant+0x562	/Users/roasbeef/gocode/src/github.com/lightningnetwork/lnd/contractcourt/channel_arbitrator.go:1512```

**Comment by Roasbeef** [comment]
New error hit in the integration tests?
```
    --- PASS: TestLightningNetworkDaemon/revoked_uncooperative_close_retribution_remote_hodl (20.67s)
	lnd_test.go:100: lnd finished with error (stderr):
		signal: killed
		
    --- FAIL: TestLightningNetworkDaemon/query_routes (26.34s)
    	lnd_test.go:69: Failed: (query routes): exited with error: 
    		*errors.errorString unable to create new nodes: context deadline exceeded
    		/home/travis/gopath/src/github.com/lightningnetwork/lnd/lnd_test.go:9057 (0xb4d430)
    			testQueryRoutes: t.Fatalf("unable to create new nodes: %v", err)
    		/home/travis/gopath/src/github.com/lightningnetwork/lnd/lnd_test.go:94 (0xb147fb)
    			(*harnessTest).RunTestCase: testCase.test(net, h)
    		/home/travis/gopath/src/github.com/lightningnetwork/lnd/lnd_test.go:9479 (0xb599a9)
    			TestLightningNetworkDaemon.func3: ht.RunTestCase(testCase, lndHarness)
    		/home/travis/.gimme/versions/go1.10.linux.amd64/src/testing/testing.go:777 (0x4ef5e0)
    			tRunner: fn(t)
    		/home/travis/.gimme/versions/go1.10.linux.amd64/src/runtime/asm_amd64.s:2361 (0x45c651)
    			goexit: BYTE	$0x90	// NOP
```

**Comment by Roasbeef** [comment]
Pulled down the logs to investigate if this is persistent.

File: htlcswitch/link.go [hunk 9]
```diff
@@ -2503,12 +2610,59 @@ func (l *channelLink) sendMalformedHTLCError(htlcIndex uint64,
...
-	log.Error(reason)
-	go l.cfg.Peer.Disconnect(reason)
+	l.errorf("Failing link: %s", reason)
+
+	// We launch a go routine to disconnect the peer and optionally force
+	// close the channel. We must do this since these call will attempt to
+	// stop the link, and wait for it to finish shutting down, and we must
+	// return from fail() to allow the link to shut down completely.
+	go func() {
+		// Send an error to the peer, why we failed the channel.
```
[the remark this answers is not in the document]
Comment (halseth) [reply htlcswitch/link.go]: The reason that is hard to do without starting a goroutine is that we must allow the `fail()` call to return for the link to shutdown (to allow the `Stop()` call triggered by the `Disconnect` to return).

We could add a call to `WaitForDisconnected` here, but I don't think it would really matter (the call to `ForceClose` would stop and wait for the disconnect to finish regardless)

**Comment by halseth** [comment]
Pushed a fix for the failing integration test. Turned out I had forgotten to shutdown the node created during the test.

File: htlcswitch/link.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (cfromknecht) [reply htlcswitch/link.go]: I'd be in favor of synchronizing this more heavily. Even if the disconnect goes through, there could still be messages queued in the mailbox that could be processed in some instances. There are places in the link where we disconnect, but don't instruct the htlc manager to exit, which seems like it could be a problem.

One way we can go about this is to use `fail` to _schedule_ a disconnect and/or force close by caching the `linkFailureError` inside the link, and then executing the bulk of this logic in a defer statement as the `htlcManager` is actually exiting.

One thing to consider, is that even after some calls to `fail` complete, the main event loop can still execute if we have messages queued up in mailboxes, event timers/block epochs fire, etc. We can prevent some of these from being delivered by nilling out channels such as `downstream` and `upstream`, stopping the packet overflow queue, etc.

Downside of this is that it could end up being a modest reworking of the link shutdown :(

[the remark this answers is not in the document]
Comment (halseth) [reply htlcswitch/link.go]: What about adding a buffered `linkFailures chan linkFailureError` to the link, that we can send on from `fail(). We can then check this channel in the `htlcManager` before it exits.

This wouldn't help with the deadlock scenario though, as we would still want the link to properly exit before we can execute the force close. What we could do to resolve this is to define a failure callback in the link config (or a channel if we don't want to introduce a new potential deadlock scenario), that is populated by the switch. We could then trigger the force close (and the disconnect) within this callback after the link is properly shut down.

Thoughts?

[the remark this answers is not in the document]
Comment (cfromknecht) [reply htlcswitch/link.go]: If the `linkFailures` channel is being read in the main event loop, there's no guarantee that the failure will be pulled off next, which may allow other updates/actions to continue before actually exiting.

IMO the easiest solutions is just to make a `linkFailure` member variable on the channel link, and just check at the beginning of each loop to see if the value is non-nil. This way, we always exit immediately after detecting that a `case` in the main event loop failed. Most of the logic currently in `fail` could be moved to the end of `htlcManager`, i.e. sending errors/force closing, to handle the failure gracefully.

We may also need to switch the ordering of force closing and disconnection, otherwise we could disconnect, and then reconnect quickly and the channel will still be open in channeldb. 

After we've broken out of the main event loop and force closed, it seems the synchronicity requirements for disconnect become less stringent. It should be safe just to do `go Disconnect()` at the end of the `htlcManager`, i think?

[the remark this answers is not in the document]
Comment (halseth) [reply htlcswitch/link.go]: Added a memeber variable as suggested, to make sure we don't process more updates after a failure is encountered.

In addition I moved the actual force close the `ChannelLinkConfig` creator (`peer`), since it has better control of the link lifecycle.

Finally, as suggested I made disconnection be done after force close is completed.

PTAL.

[the remark this answers is not in the document]
Comment (cfromknecht) [reply htlcswitch/link.go]: @halseth sweet, taking another look :)

File: peer.go [hunk 10]
```diff
@@ -454,6 +424,128 @@ func (p *peer) loadActiveChannels(chans []*channeldb.OpenChannel) error {
...
+					}, true)
+					if err != nil {
+						peerLog.Errorf("unable to "+
+							"send msg to remote "+
+							"peer: %v", err)
+					}
+				}
+
+				// Initiate disconnection.
+				p.Disconnect(linkErr)
```
Comment (cfromknecht) [inline peer.go]: Thinking about this disconnect, what if we have other channels that didn't fail? Is just removing the link sufficient?

File: peer.go [hunk 11]
```diff
@@ -454,6 +424,128 @@ func (p *peer) loadActiveChannels(chans []*channeldb.OpenChannel) error {
...
+		FeeEstimator:   p.server.cc.feeEstimator,
+		BlockEpochs:    blockEpoch,
+		PreimageCache:  p.server.witnessBeacon,
+		ChainEvents:    chainEvents,
+		UpdateContractSignals: func(signals *contractcourt.ContractSignals) error {
+			return p.server.chainArb.UpdateContractSignals(
+				*chanPoint, signals,
+			)
+		},
+		OnChannelFailure: func(chanID lnwire.ChannelID,
```
Comment (cfromknecht) [inline peer.go]: Perhaps move this into it's own method, it would improve the readability of the configuration initialization

File: peer.go [hunk 12]
```diff
@@ -454,6 +424,128 @@ func (p *peer) loadActiveChannels(chans []*channeldb.OpenChannel) error {
 	return nil
 }
 
+// addLink creates and adds a new link from the specified channel.
+func (p *peer) addLink(chanPoint *wire.OutPoint,
```
Comment (cfromknecht) [inline peer.go]: 👍

File: htlcswitch/link.go [hunk 13]
```diff
@@ -33,14 +33,93 @@ const (
...
 var (
-	// ErrInternalLinkFailure is a generic error returned to the remote
-	// party so as to obfuscate the true failure.
-	ErrInternalLinkFailure = errors.New("internal link failure")
-
 	// ErrLinkShuttingDown signals that the link is shutting down.
 	ErrLinkShuttingDown = errors.New("link shutting down")
 )
 
+// errorCode encodes the possible types of errors that will make us fail the
```
Comment (cfromknecht) [inline htlcswitch/link.go]: Maybe justified in having it's own file?

File: htlcswitch/link.go [hunk 14]
```diff
@@ -179,6 +258,16 @@ type ChannelLinkConfig struct {
 	// subscribed to new events.
 	PreimageCache contractcourt.WitnessBeacon
 
+	// OnChannelFailure is a function closure that we'll call if the
+	// channel fail for some reason. Depending on the severity of the
```
Comment (cfromknecht) [inline htlcswitch/link.go]: s/fail/failed/

File: htlcswitch/link_test.go [hunk 15]
```diff
@@ -4256,3 +4295,223 @@ func TestChannelLinkWaitForRevocation(t *testing.T) {
...
+}
+
+func (*mockPackager) AckSettleFails(tx *bolt.Tx,
+	settleFailRefs ...channeldb.SettleFailRef) error {
+	return nil
+}
+
+// TestChannelLinkFail tests that we will fail the channel, and force close the
+// channel in certain situations.
+func TestChannelLinkFail(t *testing.T) {
```
Comment (cfromknecht) [inline htlcswitch/link_test.go]: Super cool test :)

File: htlcswitch/link.go [hunk 16]
```diff
@@ -2503,12 +2613,16 @@ func (l *channelLink) sendMalformedHTLCError(htlcIndex uint64,
...
+// to the OnChannelFailure closure, in order for it to determine if we should
+// force close the channel, and if we should send an error message to the
+// remote peer.
+func (l *channelLink) fail(linkErr LinkFailureError,
+	format string, a ...interface{}) {
 	reason := errors.Errorf(format, a...)
-	log.Error(reason)
-	go l.cfg.Peer.Disconnect(reason)
+	l.errorf("Failing link: %s", reason)
+	l.cfg.OnChannelFailure(l.ChanID(), l.ShortChanID(), linkErr)
```
Comment (Roasbeef) [inline htlcswitch/link.go]: If the `Disconnect` method on the peer is no longer used, we may want to consider removing it all together. This isn't a blocker for this PR though, and can be done at a later time.

File: htlcswitch/link.go [hunk 17]
```diff
@@ -342,6 +342,10 @@ type channelLink struct {
 	started  int32
 	shutdown int32
 
+	// failed should be set to true in case a link error happens, making
```
Comment (Roasbeef) [inline htlcswitch/link.go]: 👍

File: peer.go (hunk 11 above)
Comment (Roasbeef) [reply peer.go]: Could as easily just be raised to a local closure right above the initialization of the config itself.

File: htlcswitch/link.go [hunk 18]
```diff
@@ -2503,12 +2634,20 @@ func (l *channelLink) sendMalformedHTLCError(htlcIndex uint64,
...
+// fail is a function which is used to encapsulate the action necessary for
+// properly failing the link. It takes a LinkFailureError, which will be passed
+// to the OnChannelFailure closure, in order for it to determine if we should
+// force close the channel, and if we should send an error message to the
+// remote peer.
+func (l *channelLink) fail(linkErr LinkFailureError,
+	format string, a ...interface{}) {
 	reason := errors.Errorf(format, a...)
-	log.Error(reason)
-	go l.cfg.Peer.Disconnect(reason)
```
Comment (Roasbeef) [inline htlcswitch/link.go]: If the `Disconnect` method is no longer used anywhere, then we may want to consider removing it from the interface all together. This isn't a blocker for this PR, though and can be added as a follow up PR.

File: peer.go (hunk 10 above)
Comment (Roasbeef) [reply peer.go]: FWIW the existing behavior would _always_ remove the connection regardless of if there were other existing channels w/ that peer.

Comment (halseth) [reply peer.go]: Added TODO, in favour of keeping the existing behaviour for now.

File: peer.go (hunk 11 above)
Comment (halseth) [reply peer.go]: I like the local closure suggestion, as it won't clutter the global surface :)

File: htlcswitch/link.go (hunk 16 above)
Comment (halseth) [reply htlcswitch/link.go]: Removed.

**Comment by halseth** [comment]
Thanks for the great suggestions! I did a scan to see if there was more places where we would fail the channel, but not return properly. Also made sure that the `fail` method will not attempt to fail the link a second time if the `failed` variable is set.  Should be fixed as part of one of the fixups.

File: htlcswitch/link.go (hunk 13 above)
Comment (halseth) [reply htlcswitch/link.go]: Done :)

File: htlcswitch/link.go [hunk 19]
```diff
@@ -1345,16 +1374,26 @@ func (l *channelLink) handleUpstreamMsg(msg lnwire.Message) {
...
 		}
 
 		l.processRemoteSettleFails(fwdPkg, settleFails)
-
 		needUpdate := l.processRemoteAdds(fwdPkg, adds)
+
+		// If the link failed during processing the adds, we must
+		// return to ensure we won't attempted to update the state
+		// further.
+		if l.failed {
```
Comment (cfromknecht) [inline htlcswitch/link.go]: This one isn't strictly necessary because we always return `false` when `processRemoteAdds` fails, but still a good sanity check all together 👍

File: htlcswitch/link.go [hunk 20]
```diff
@@ -2503,12 +2560,26 @@ func (l *channelLink) sendMalformedHTLCError(htlcIndex uint64,
...
+func (l *channelLink) fail(linkErr LinkFailureError,
+	format string, a ...interface{}) {
 	reason := errors.Errorf(format, a...)
-	log.Error(reason)
-	go l.cfg.Peer.Disconnect(reason)
+	l.errorf("Failing link: %s", reason)
+
+	// Return if we have already notified about a failure.
+	if l.failed {
+		l.warnf("Already failed")
```
Comment (cfromknecht) [inline htlcswitch/link.go]: Might be worthwhile to log the reason here as well.

File: htlcswitch/link.go [hunk 21]
```diff
@@ -2503,12 +2560,26 @@ func (l *channelLink) sendMalformedHTLCError(htlcIndex uint64,
...
+// properly failing the link. It takes a LinkFailureError, which will be passed
+// to the OnChannelFailure closure, in order for it to determine if we should
+// force close the channel, and if we should send an error message to the
+// remote peer.
+func (l *channelLink) fail(linkErr LinkFailureError,
+	format string, a ...interface{}) {
 	reason := errors.Errorf(format, a...)
-	log.Error(reason)
-	go l.cfg.Peer.Disconnect(reason)
+	l.errorf("Failing link: %s", reason)
```
Comment (cfromknecht) [inline htlcswitch/link.go]: With the new check below, it seems this statement should be moved after the check, so that this log only happens on the first failure.

File: htlcswitch/link.go [hunk 22]
```diff
@@ -694,6 +699,13 @@ func (l *channelLink) resolveFwdPkg(fwdPkg *channeldb.FwdPkg) (bool, error) {
 			fwdPkg.Source, fwdPkg.Height, fwdPkg.Adds,
 		)
 		needUpdate = l.processRemoteAdds(fwdPkg, adds)
+
+		// If the link failed during processing the adds, we must
+		// return to ensure we won't attempted to update the state
+		// further.
+		if l.failed {
+			return false, fmt.Errorf("link failed")
```
Comment (cfromknecht) [inline htlcswitch/link.go]: Would prefer a more descriptive error here, as we may log it in the `htlcManager`'s call to `fail`.

File: peer.go [hunk 23]
```diff
@@ -454,6 +424,130 @@ func (p *peer) loadActiveChannels(chans []*channeldb.OpenChannel) error {
...
+					shortChanID, err)
+			}
+
+			// If the error encountered was severe enough, we'll
+			// now force close the channel.
+			if linkErr.ForceClose {
+				peerLog.Warnf("Force closing link(%v)",
+					shortChanID)
+
+				closeTx, err := p.server.chainArb.ForceCloseContract(*chanPoint)
```
Comment (cfromknecht) [inline peer.go]: Might we want this to execute before spawning the goroutine, to prevent any possibility of the link being readded to the switch before this executes? I think there could be interleaving that could permit this if the remote party disconnects and reconnects super quick, and the db still shows the channel as open.

Granted, it would likely be difficult in practice :P

Would this cause any races you're aware of with the contract court?

File: peer.go [hunk 24]
```diff
@@ -454,6 +424,130 @@ func (p *peer) loadActiveChannels(chans []*channeldb.OpenChannel) error {
...
+		// must let OnChannelFailure return in order for the link to
+		// completely stop in the call to RemoveLink.
+		p.wg.Add(1)
+		go func() {
+			defer p.wg.Done()
+
+			// We begin by removing the link from the switch, such
+			// that it won't be attempted used for any more
+			// updates.
+			err := p.server.htlcSwitch.RemoveLink(chanID)
```
Comment (cfromknecht) [inline peer.go]: All links for this peer will cleaned up by the `peerTerminationWatcher` upon detecting a disconnect. If the link has already broken its event loop, why is this needed?

File: peer.go [hunk 25]
```diff
@@ -454,6 +424,130 @@ func (p *peer) loadActiveChannels(chans []*channeldb.OpenChannel) error {
...
+		BatchSize:    10,
+		UnsafeReplay: cfg.UnsafeReplay,
+	}
+	link := htlcswitch.NewChannelLink(linkCfg, lnChan,
+		uint32(currentHeight))
+
+	// With the channel link created, we'll now notify the htlc switch so
+	// this channel can be used to dispatch local payments and also
+	// passively forward payments.
+	if err := p.server.htlcSwitch.AddLink(link); err != nil {
```
Comment (cfromknecht) [inline peer.go]: Return error directly? 🙃

File: peer.go (hunk 24 above)
Comment (halseth) [reply peer.go]: It was meant to have the effect of making sure it is not attempted to be used for routing, since we know at this point that the link won't be useful. 

Not sure how much of an effect it will have though, but my assumption was that it might take a while before the `peerTerminationWatcher` removes it?

Comment (halseth) [reply peer.go]: Also note that the call to `ForceClose` below will also call `RemoveLink`, but that is only in the few force close cases, ofc.

Comment (cfromknecht) [reply peer.go]: Ahh I see, yeah sounds like it can't hurt!

File: peer.go (hunk 23 above)
Comment (halseth) [reply peer.go]: This was attempted at first, but it will cause a deadlock as the `ChannelArbitrator` will call `RemoveLink`, which in turns call `link.Stop()`.

Maybe this won't be a problem since we force close before we disconnect the peer?

File: peer.go (hunk 25 above)
Comment (halseth) [reply peer.go]: Yeeeeeh

File: peer.go (hunk 23 above)
Comment (cfromknecht) [reply peer.go]: Well it's safe if we assume that we're the only ones who can disconnect, but the issue i'm thinking of happens when the remote peer disconnects, which we can't control

Comment (cfromknecht) [reply peer.go]: Alright, verdict on this was delay the untangling of switch/chainarb/peer life cycle around force close to a later PR. We will circle back and lock everything down in a follow up.

**Comment by halseth** [comment]
Squashed and rebase.
